Don't use RefPtr<MediaElementSession> in HTMLMediaElement::virtualHasPendingActivity

fujii opened the pull request:
#### e2f9f386c2f59d4947e9cebebef32b11c9845289
<pre>
Don&apos;t use RefPtr&lt;MediaElementSession&gt; in HTMLMediaElement::virtualHasPendingActivity
<a href="https://bugs.webkit.org/show_bug.cgi?id=291413">https://bugs.webkit.org/show_bug.cgi?id=291413</a>

Reviewed by Chris Dumez.

After <a href="https://commits.webkit.org/293547@main">https://commits.webkit.org/293547@main</a> added RefPtr&lt;MediaElementSession&gt; in
HTMLMediaElement::virtualHasPendingActivity, a threading assertion was
failing while destructing objects related to MediaElementSession.

&gt; ASSERTION FAILED: Unsafe to ref/deref from different threads
&gt; m_isOwnedByMainThread == isMainThread()

HTMLMediaElement::virtualHasPendingActivity could be called on a
non-main thread. We shouldn&apos;t use RefPtr&lt;MediaElementSession&gt; in it.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::virtualHasPendingActivity const):
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::mediaSessionIfExists const):

Canonical link: <a href="https://commits.webkit.org/293648@main">https://commits.webkit.org/293648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69cc3ae68a95719d029c6034f8f010950c2d2b0d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99484 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9387 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104615 "Hash 69cc3ae6 for PR 43990 does not build (failure)") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50086 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101525 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19422 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27567 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/104615 "Hash 69cc3ae6 for PR 43990 does not build (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32823 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102491 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14788 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89829 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/104615 "Hash 69cc3ae6 for PR 43990 does not build (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14585 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7813 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49445 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84514 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7900 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106973 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26598 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19411 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84681 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26960 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86033 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84199 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28874 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6569 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20376 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16195 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26538 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31739 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26358 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29671 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27925 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6788,7 +6788,7 @@ bool HTMLMediaElement::virtualHasPendingActivity() const
             return true;
 
-        RefPtr mediaSession = this->mediaSessionIfExists();
+        SUPPRESS_UNCOUNTED_LOCAL auto* mediaSession = this->mediaSessionIfExists();
```
Please a comment saying this function could be called on a non-main thread.

fujii replied:
this function already has a comment above.

fujii replied:
https://github.com/WebKit/WebKit/blob/39d4b41e86f73983c9cc8bcdce7ac9dee35ae94c/Source/WebCore/html/HTMLMediaElement.cpp#L6778

fujii replied:
Fixed.

fujii committed:
Don't use RefPtr<MediaElementSession> in HTMLMediaElement::virtualHasPendingActivity
https://bugs.webkit.org/show_bug.cgi?id=291413

Reviewed by Chris Dumez.

After https://commits.webkit.org/293547@main added RefPtr<MediaElementSession> in
HTMLMediaElement::virtualHasPendingActivity, a threading assertion was
failing while destructing objects related to MediaElementSession.

> ASSERTION FAILED: Unsafe to ref/deref from different threads
> m_isOwnedByMainThread == isMainThread()

HTMLMediaElement::virtualHasPendingActivity could be called on a
non-main thread. We shouldn't use RefPtr<MediaElementSession> in it.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::virtualHasPendingActivity const):
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::mediaSessionIfExists const):

Canonical link: https://commits.webkit.org/293648@main
